Email option to embed images as base64 instead of link

sommerf-lf opened the pull request:
ref: #15081
ref: #14037

Documentation: https://gitea.com/gitea/docs/pulls/69

# Example
Content:
![image](https://github.com/user-attachments/assets/e73ebfbe-e329-40f6-9c4a-f73832bbb181)
Result in Email:
![image](https://github.com/user-attachments/assets/55b7019f-e17a-46c3-a374-3b4769d5c2d6)
Result with source code:
(first image is external image, 2nd is now embedded)
![image](https://github.com/user-attachments/assets/8e2804a1-580f-4a69-adcb-cc5d16f7da81)

sommerf-lf commented:
Please have a look at this and let me know if I did some structural things wrong, as I have zero experience with Go but came across the mentioned issued myself and tried to add the optional setting to fix them.

lunny commented:
Could you have some tests?

sommerf-lf commented:
> Could you have some tests?

I'll have a look. But I'm not quite sure as to what the focus of these test should be, as outgoing emails aren't tested anywhere. I could simply test my own additions whithout any actual email integration (the parsing of the html and replacement of the img tags and whether other image tags stay the same). Sounds fine?

lunny commented:
> > Could you have some tests?
> 
> I'll have a look. But I'm not quite sure as to what the focus of these test should be, as outgoing emails aren't tested anywhere. I could simply test my own additions whithout any actual email integration (the parsing of the html and replacement of the img tags and whether other image tags stay the same). Sounds fine?

Yes. Thank you.

sommerf-lf commented:
I added test to this. however I can't integrate them into services/mailer/mail_test.go because I need the testing environment which would form an import cycle. Now I exported the MailCommentContext and my newly added functions. Let me know if there is a way around this, but I see this as nothing big as these functions are not context specific. I'll squash the commits after you reviewed it

sommerf-lf commented:
Still fails because I didn't account for using minio. Will rework...

yp05327 commented on services/mailer/mail.go:
```diff
@@ -363,6 +377,81 @@ func composeIssueCommentMessages(ctx *mailCommentContext, lang string, recipient
+							continue
+						}
+						log.Trace("Old value of src attribute: %s, new value (first 100 characters): %s", attr.Val, dataURI[:100])
+						n.Attr[i].Val = dataURI
```
We can have a `break` below?

yp05327 commented on tests/integration/email_embed_b64_images_test.go:
```diff
@@ -0,0 +1,71 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
```
2024?

yp05327 commented:
https://github.blog/changelog/2023-05-09-more-secure-private-attachments/
The discussion in the issue is outdated. See the link above.
Personally, I will not suggest do this.

sommerf-lf commented:
> https://github.blog/changelog/2023-05-09-more-secure-private-attachments/ The discussion in the issue is outdated. See the link above. Personally, I don't suggest to do this.

I see your/their point. Even though it is not the exact same situation, the motivation for doing this is what our issue (that we are trying to resolve) is. For Github I see this as a more sensitive issue, as it is one instance that is publically hosted. With this change we are not allowing any urls to access attachments on gitea repos without authentification (no chance that this change will open any possibility of brute-forcing attachments), but instead we are "trusting" the email chain. (and for images only)
As I implemented this as an opt-in option, I suggest that I add more info (one sentence about the security implications) to the documentation and the `app.example.ini` to warn about this possible concern.
Does this resolve your concerns?

yp05327 commented:
I see. You added an option. It is acceptable.
But the default value is `false` in code and `true` in app.ini
![image](https://github.com/user-attachments/assets/43600b88-689e-4de9-ba31-9e54d973f9b7)
![image](https://github.com/user-attachments/assets/01db4f95-aade-4de4-bd63-6753f703336a)

sommerf-lf commented:
Should I add an info to the documentation of the implication of email trust with this feature?

wxiaoguang commented:
Before reviewing some nits, actually there are some more important problems in my mind.

1. Is it safe to make `AttachmentSrcToBase64DataURI` read attachments directly? It means that it could read **any** attachment without permission check? If I remember correctly, there are some permission checks for attachment access.
2. If an attacker use `<img src="/attachments/{a-big-release-file}>`, would it bloat Gitea's memory and make it OOM?

yp05327 commented:
> Is it safe to make AttachmentSrcToBase64DataURI read attachments directly? It means that it could read any attachment without permission check? If I remember correctly, there are some permission checks for attachment access.

I think this PR is based on your comment here: https://github.com/go-gitea/gitea/issues/15081#issuecomment-914063376
But as I mentioned above https://github.com/go-gitea/gitea/pull/32061#issuecomment-2375547542, GitHub has changed it.
So I think this concern is necessary.

And as https://github.com/go-gitea/gitea/pull/32061#issuecomment-2384137381 explained, by default, this feature will be turn off, and only for these internal instances with special needs. So from my side, by default, this feature will not work, so it will not break the security of existing instances. We can just mention the risk in docs and `app.simple.ini`, if user want to use it, so maybe it is ok.
But, at last, I think we need TOC's decision.

sommerf-lf commented:
> Before reviewing some nits, actually there are some more important problems in my mind.
> 
> 1. Is it safe to make `AttachmentSrcToBase64DataURI` read attachments directly? It means that it could read **any** attachment without permission check? If I remember correctly, there are some permission checks for attachment access.
> 2. If an attacker use `<img src="/attachments/{a-big-release-file}>`, would it bloat Gitea's memory and make it OOM?

I addressed both your points:
1. checked if COMMENT CREATOR has rights to view the attachment (still only works for images!), as the sender is unknown in the context of the mail (bulk mails). this is not ideal but a good check, as this prevent data loss from brute-forcing other attachments as a comment creator
2. added max size to app.ini (default 5242880 = 5mb)

sommerf-lf commented:
> > Is it safe to make AttachmentSrcToBase64DataURI read attachments directly? It means that it could read any attachment without permission check? If I remember correctly, there are some permission checks for attachment access.
> 
> I think this PR is based on your comment here: [#15081 (comment)](https://github.com/go-gitea/gitea/issues/15081#issuecomment-914063376) But as I mentioned above [#32061 (comment)](https://github.com/go-gitea/gitea/pull/32061#issuecomment-2375547542), GitHub has changed it. So I think this concern is necessary.
> 
> And as [#32061 (comment)](https://github.com/go-gitea/gitea/pull/32061#issuecomment-2384137381) explained, by default, this feature will be turn off, and only for these internal instances with special needs. So from my side, by default, this feature will not work, so it will not break the security of existing instances. We can just mention the risk in docs and `app.example.ini`, if user want to use it, so maybe it is ok. But, at last, I think we need TOC's decision.

i will address the docu and `app.example.ini` later this week with a description of the limitation and followed by a warning of it's implication.

Sounds good to me so far.

wxiaoguang commented on custom/conf/app.example.ini:
```diff
@@ -1704,6 +1704,12 @@ LEVEL = Info
+;BASE64_EMBED_IMAGES = false
+;;
+;; The maximum size of an image attachment to be embedded in the email.
+;BASE64_EMBED_IMAGES_MAX_SIZE = 5242880
```
I guess we do not need to introduce a new option here. There are already too many config options.

A default value like 5M should be good enough for all users. If there would be some special requirements, it won't be late to introduce an option the.

And one more thing, limiting a single file seems not enough. There could be 100 files * 5M , I guess nobody would really like to receive an email with 500M size.

hiifong replied:
> Maybe you can just resize the bigger images to fixed width/height before base64 encode so they will take up much less space?

I think we have to compress the image before base64 it

hiifong replied:
As far as I know, most email providers limit the size of the entire email to 10-30M, but after base64, the attachment will expand by about 1.5 times, and the actual size of the attachment that can be added will be further reduced.

sommerf-lf replied:
I reworked this to sum up the attachments in an email (and then stop if the threshold would be reached). New config is `BASE64_EMBED_IMAGES_MAX_SIZE_PER_EMAIL`.
I suggest (and did in my commit) to set this by default to 9.5 MB (`BASE64_EMBED_IMAGES` is still disabled by default), which results in the whole email beeing less than 10 MB, even with a long mail body. This should then be accepted by all standard email programs' configs.
For compression, I don't see a need for that, as it would further complicate the function and having a look at our company data most images are below 200 KB.

wxiaoguang commented on services/mailer/mail.go:
```diff
@@ -363,6 +379,102 @@ func composeIssueCommentMessages(ctx *mailCommentContext, lang string, recipient
+	}
+	defer fr.Close()
+
+	content, err := io.ReadAll(fr)
```
It should avoid using `ReadAll` when the size is not under control.

There are some limited readers for this case.

sommerf-lf replied:
True, this might in edge cases result in unforseen situations. I fixed this my reading at max the size configured in the config: `BASE64_EMBED_IMAGES_MAX_SIZE_PER_EMAIL`

wxiaoguang commented on tests/testdata/data/attachments/1/b/1b267670-1793-4cd0-abc1-449269b7cff9:
Do we really the 1.4KB size to test? 

* A valid JPEG can be as small as 107 bytes
* The smallest PNG file is 67 bytes

Since the image file is only used by this single test, I think a sample image file (and database fixture) could be generated on the fly?

sommerf-lf replied:
done. Still 270 Bytes, but it is a valid png, not just an "empty" one

wxiaoguang commented on tests/integration/email_embed_b64_images_test.go:
```diff
@@ -0,0 +1,71 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestEmailEmbedBase64Images(t *testing.T) {
```
If I understand correctly, this test could be a "unit test" (in `services/mailer/mail_xxx_test.go`), but not a "integration test"? Then no need to expose `mailCommentContext`?

sommerf-lf replied:
That's correct. done

a1012112796 commented:
A little wander about whether embedding images in emails will makes them more likely to be taged as spam?

wxiaoguang commented:
> A little wander about whether embedding images in emails will makes them more likely to be taged as spam?

It is optional and disabled by default, users could have their own choice.

KazzmanK commented:
This is going to be a killer-feature for us, please, do not postpone it to next releases.

wxiaoguang commented:
I will try to address the concerns above if I have time.

sommerf-lf commented:
Hey everyone,

sorry for the delay, life got a bit hectic with university courses and some extra work projects. I’ve reworked the PR to address your concerns, including:

- Reduced Test Image Size
The newly added test image is now only about 270 bytes (instead of 1.4 KB).
- Integrated Tests Into mail_test.go
I’ve moved the tests into services/mailer/mail_test.go, so there’s no longer a need to expose mailCommentContext.
- Additional Tests
I added more tests, including one that integrates a mocked/fake email body for better coverage.
- New Mail Size Setting
I replaced BASE64_EMBED_IMAGES_MAX_SIZE with BASE64_EMBED_IMAGES_MAX_SIZE_PER_EMAIL, defaulting to 9.5 MB. This addresses overall email size concerns. (the whole feature is still disabled by default BASE64_EMBED_IMAGES = false)
- Use of LimitedReader
When reading and embedding images, I now use a LimitedReader based on the new max-size-per-email setting. Any file exceeding this limit is ignored to prevent oversize issues.


_However I am having trouble to get the new test to run in the pipeline. They run perfectly like this: `/usr/local/go/bin/go test -c -tags sqlite,sqlite_unlock_notify -o /mnt/c/Users/sommerf/GolandProjects/gitea/gitea code.gitea.io/gitea/services/mailer`, but when calling it with make (as the pipeline does: `GO_TEST_PACKAGES='code.gitea.io/gitea/services/mailer' TESTFLAGS='-run=^TestEmbedBase64ImagesInEmail$' make unit-test-coverage`) the /tmp/go-build3776050572 directory hasn't got the attachments, which results in the test failing. **Help would be greatly appreciated**, as I am unfortunately way over my head in debugging the "go test environment"._

This is actually my first time using Go, so I’m really grateful for the feedback and support. Additionally, I’ll update the documentation for BASE64_EMBED_IMAGES_MAX_SIZE_PER_EMAIL in the docs repo. If you see anything else that needs adjusting or you’d like these changes handled differently, just let me know and I’ll see when I can work on it.

Thanks for your patience and feedback!

– Felix

wxiaoguang commented:
Thank you very much! Will take a look and fix the tests later.

wxiaoguang commented:
I can see there is a new related PR " make private image in email can be read by security link #33757 "

If we could have #33757, do we still need to embed the images as base64 in email body?

I can see the Pros of #33757, including no need extra config option, small email body, while no Cons.

KazzmanK commented:
> do we still need to embed the images as base64

These two features are not mutually exclusive. In our case Gitea server is not accessible from internet, so having links of any sort in emails makes no use.
Instead, base64 makes images accessible.

wxiaoguang commented:
Will do some refactoring first before continuing this. Thank you for your patience.

wxiaoguang commented:
Made some changes:

* Only keep one "EMBED_ATTACHMENT_IMAGES" config option, the size limit is hard-coded to 9MB. I think this limit is good enough to most users, and avoid bloating Gitea's config system too much.
    * If there are special requests, there is always a chance to add the "limit" config option back.
* Instead of "guessing" the "attachments" path, use a strict parser to parse Gitea's route path
